Content.Client/StatusIcon/StatusIconSystem.cs

```diff
@@ -54,10 +60,34 @@ public List<StatusIconData> GetStatusIcons(EntityUid uid, MetaDataComponent? met
         if (meta.EntityLifeStage >= EntityLifeStage.Terminating)
             return list;
 
-        var inContainer = (meta.Flags & MetaDataFlags.InContainer) != 0;
-        var ev = new GetStatusIconsEvent(list, inContainer);
+        var ev = new GetStatusIconsEvent(list);
         RaiseLocalEvent(uid, ref ev);
         return ev.StatusIcons;
     }
-}
 
+    /// <summary>
+    /// For overlay to check if an entity can be seen.
+    /// </summary>
+    public bool IsVisible(Entity<MetaDataComponent> ent, StatusIconData data)
+    {
+        var viewer = _playerManager.LocalSession?.AttachedEntity;
+
+        // Always show our icons to our entity
+        if (viewer == ent.Owner)
+            return true;
+
+        if (data.VisibleToGhosts && HasComp<GhostComponent>(viewer))
+            return true;
+
+        if (data.HideInContainer && (ent.Comp.Flags & MetaDataFlags.InContainer) != 0)
+            return false;
+
+        if (data.HideOnStealth && TryComp<StealthComponent>(ent, out var stealth) && stealth.Enabled)
+            return false;
```

Comment:
Shouldn't this also exempt the stealth owner?

Reply:
i mean, this wont ever run for the owner cuz of the `viewer == ent.Owner return true` check above, right?

Reply:
Ah nvm I didn't see that